### PR TITLE
Big old refactor

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Refactor parser internals to eliminate duplication and improve modularity by introducing a shared ParserCore and dedicated content builders, while adding a new JsonNumber::from_slice API and updating tests accordingly.

New Features:
- Add JsonNumber::from_slice for parsing JSON numbers directly from byte slices.

Enhancements:
- Extract shared parsing logic into a new ParserCore with a unified event processing loop.
- Introduce SliceContentBuilder and StreamContentBuilder to isolate parser-specific input reading and content extraction.
- Refactor SliceParser and StreamParser to delegate core parsing steps to ParserCore and the new builders.
- Update escape_processor to return UTF-8 bytes by value and add a Default impl for UnicodeEscapeCollector.

Tests:
- Add unit tests for JsonNumber::from_slice and update existing event processor and parser tests to reflect the new abstractions.

## Summary by CodeRabbit

* **New Features**
  * Introduced unified core parsing logic for both slice-based and stream-based JSON parsing, improving consistency and maintainability.
  * Added new content builders for efficient and flexible JSON content extraction from slices and streams.
  * Added a public method to create JSON numbers directly from byte slices, centralizing number parsing logic.

* **Refactor**
  * Simplified and centralized event processing and escape sequence handling, reducing code duplication.
  * Streamlined internal parser state management for improved performance and reliability.
  * Removed redundant traits and free functions related to number extraction and escape handling.
  * Consolidated parsing responsibilities into shared core and content builder components.
  * Refactored slice and stream parsers to delegate parsing and escape handling to core and builder abstractions.

* **Bug Fixes**
  * Enhanced safety in buffer operations and number parsing, reducing risk of panics and parsing errors.
  * Replaced unchecked arithmetic and panicking copy operations with safe alternatives.

* **Documentation**
  * Expanded and updated unit tests to cover new parsing behaviors and edge cases.

### CODE REVIEW

File: picojson/src/event_processor.rs [hunk 1]
```diff
@@ -5,33 +5,165 @@
...
+                    | EventToken::EscapeCarriageReturn
+                    | EventToken::EscapeTab),
+                ) if escape_timing == EscapeTiming::OnEnd => {
+                    // For StreamParser, the escape ends here.
+                    provider.process_simple_escape_event(&escape_token)?;
+                    self.in_escape_sequence = false;
+                }
+                _ => {
+                    // All other events continue to next iteration
+                }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/event_processor.rs:146]: Nah

File: picojson/src/json_number.rs [hunk 2]
```diff
@@ -56,6 +56,37 @@ pub enum JsonNumber<'a, 'b> {
...
+            parse_integer(bytes)
+        } else {
+            #[cfg(feature = "float")]
+            {
+                parse_float(bytes)
+            }
+            #[cfg(not(feature = "float"))]
+            {
+                parse_float(bytes)?
+            }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/json_number.rs:81]: It's correct

File: picojson/src/event_processor.rs [hunk 3]
```diff
@@ -5,33 +5,165 @@
...
+            }
 
-    /// Begin escape sequence processing (lifecycle method with default no-op implementation)
-    /// Called when escape sequence processing begins (e.g., on Begin(EscapeSequence))
-    fn begin_escape_sequence(&mut self) -> Result<(), crate::ParseError> {
-        Ok(())
+            let taken_event = take_first_event(&mut self.parser_state.evts);
+            let Some(taken) = taken_event else {
+                return Err(UnexpectedState::StateMismatch.into());
+            };
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/event_processor.rs:83]: it's correct
